# **CLUSTERING NOTEBOOK**

## Libraries

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from umap.umap_ import UMAP
from sklearn.metrics import silhouette_samples, silhouette_score
import matplotlib.cm as cm
from sklearn.decomposition import PCA
import plotly.express as px
%load_ext autoreload
%autoreload 2

In [2]:
import sys
import os

# Get the absolute path of the source_code folder
source_code_path = os.path.abspath('../source_code')

# Add the source_code folder to sys.path
if source_code_path not in sys.path:
    sys.path.append(source_code_path)

from utils import *
from preprocessing import *
from clustering import *
from analysis import *

# Get the absolute path of the source_code folder
clustering_outputs_path = os.path.abspath('../clustering_outputs')

# Add the source_code folder to sys.path
if clustering_outputs_path not in sys.path:
    sys.path.append(clustering_outputs_path)



## Load Data

In [3]:
customer_info = load_dataset('../customer_info.csv')

In [4]:
#customer_info.columns

In [5]:
# Calling the clusters
final_clustering_notcaled=load_dataset('../clustering_outputs/final_clustering_notscaled.csv')
final_clustering = load_dataset('../clustering_outputs/final_clustering.csv')

In [6]:
final_clustering.cluster.unique()

array([0, 1, 2, 3, 4, 5, 6, 7, 9, 8], dtype=int64)

In [128]:
customer_basket = load_dataset('../customer_basket.csv')

## Analysis

In [7]:
vars, cluster_avg, database_avg=compute_average_multi(hierarchical8_scaled, 'cluster')

In [8]:
plot_all_clusters_profile_plotly(vars, cluster_avg, database_avg)

## Association Rules

In [129]:
df_items, items_list = transform_dataset(customer_basket, final_clustering,8)

In [130]:
df_items, items_list

(       airpods  almonds  antioxydant juice  asparagus  avocado  babies food  \
 0        False    False              False      False    False        False   
 1        False    False              False      False    False        False   
 2        False    False              False       True     True        False   
 3        False    False              False      False     True        False   
 4        False    False              False      False    False        False   
 ...        ...      ...                ...        ...      ...          ...   
 13435    False    False              False      False    False        False   
 13436     True    False              False      False     True        False   
 13437    False    False              False      False    False        False   
 13438    False    False              False      False    False        False   
 13439    False    False              False      False    False        False   
 
        bacon  barbecue sauce   beer  

In [131]:
frequent_itemsets, rules = apriori_algorithm(df_items, min_support=0.15, metric='confidence', confidence_threshold = 0.5)

In [132]:
rules = rules.sort_values('lift', ascending=False)
rules

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski
21,(spaghetti),"(fresh tuna, champagne)",0.335789,0.448065,0.180134,0.536450,1.197259,1.0,0.029679,1.190669,0.248051,0.298373,0.160136,0.469238
18,(frozen smoothie),"(fresh tuna, champagne)",0.285193,0.448065,0.152158,0.533525,1.190729,1.0,0.024372,1.183202,0.224087,0.261844,0.154836,0.436556
20,"(spaghetti, champagne)",(fresh tuna),0.281250,0.542188,0.180134,0.640476,1.181282,1.0,0.027644,1.273386,0.213512,0.280014,0.214692,0.486356
12,"(fresh tuna, champagne)",(bluetooth headphones),0.448065,0.427381,0.225893,0.504151,1.179630,1.0,0.034398,1.154826,0.275896,0.347766,0.134069,0.516351
15,(bluetooth headphones),"(fresh tuna, champagne)",0.427381,0.448065,0.225893,0.528552,1.179630,1.0,0.034398,1.170721,0.265930,0.347766,0.145825,0.516351
17,"(champagne, frozen smoothie)",(fresh tuna),0.239137,0.542188,0.152158,0.636279,1.173540,1.0,0.022501,1.258691,0.194355,0.241840,0.205524,0.458458
14,"(bluetooth headphones, champagne)",(fresh tuna),0.355655,0.542188,0.225893,0.635146,1.171452,1.0,0.033061,1.254784,0.227143,0.336175,0.203050,0.525889
11,(spaghetti),(fresh tuna),0.335789,0.542188,0.208631,0.621316,1.145943,1.0,0.026571,1.208957,0.191741,0.311694,0.172841,0.503056
19,"(fresh tuna, spaghetti)",(champagne),0.208631,0.753497,0.180134,0.863409,1.145870,1.0,0.022931,1.804685,0.160861,0.230352,0.445887,0.551237
10,(frozen smoothie),(fresh tuna),0.285193,0.542188,0.176860,0.620141,1.143776,1.0,0.022232,1.205217,0.175856,0.271875,0.170274,0.473169


In [127]:
rules = rules[(rules['support'] > 0.1) & (rules['confidence'] > 0.5) & (rules['lift'] >= 1.2)]
rules.sort_values(by = 'confidence', ascending=False)

,antecedents,consequents,antecedent support,consequent support,support,confidence,lift,representativity,leverage,conviction,zhangs_metric,jaccard,certainty,kulczynski


Assosiation Rules

Cluster 0:
(pasta, meatballs) (oil)

- Confidence: 88.9% of those who bought pasta and meatballs also bought oil
- Support: This rule appears in 15% of all transactions in the cluster 
- Lift:The likelihood of buying oil increases by 1.2 if pasta and meatballs are bought.

Cluster 1:
(cauliflower, carrots)	(tomatoes, asparagus)

- Confidence: 79.2% of those who bought cauliflower and carrots also bought tomatoes and asparagus.
- Support: This rule appears in 8% of all transactions in the cluster 
- Lift:The likelihood of buying tomatoes and asparagus increases by 1.26 if cauliflower and carrots are bought.

Cluster 2:
(minecraft, cooking oil)	(babies food, ratchet & clank)

- Confidence: 55.8% of those who bought minecraft and cooking oil also bought babies food, ratchet and clank
- Support: This rule appears in 9.2% of all transactions in the cluster 
- Lift:The likelihood of buying babies food, ratchet and clank increases by 1.2 if minecraft and cooking oil are bought.

CLuster 3:
(pancakes)	(energy drink)

- Confidence: 79.7% of those who bought pancakes also bought energy drink
- Support: This rule appears in 28.6% of all transactions in the cluster 
- Lift:The likelihood of buying energy drink increases by 1.62 if pancakes are bought.

CLuster 4:
(cake, cooking oil)	(oil, tea)	

- Confidence: 51.6% of those who bought cake and cooking oil also bought oil and tea.
- Support: This rule appears in 12% of all transactions in the cluster 
- Lift:The likelihood of buying oil and tea increases by 1.22 if cake and cooking oil are bought.

CLuster 5:
(salt)	(beer)	0.341256	0.655605	0.280717	0.822602	1.254721
(white wine)	(beer)	0.534081	0.655605	0.395067	0.739715	1.128292

- Confidence: 88.9% of those who bought pasta and meatballs also bought oil
- Support: This rule appears in 15% of all transactions in the cluster 
- Lift:The likelihood of buying oil increases by 1.2 if pasta and meatballs are bought.

Cluster 6:
(gums)	(oil)	0.268779	0.723556	0.223812	0.832700	1.150843
(candy bars, cooking oil)	(oil)	0.175779	0.723556	0.153807	0.875000	1.209304

- Confidence: 88.9% of those who bought pasta and meatballs also bought oil
- Support: This rule appears in 15% of all transactions in the cluster 
- Lift:The likelihood of buying oil increases by 1.2 if pasta and meatballs are bought.

Cluster 7:
(cake, cooking oil)	(oil)	0.258852	0.770106	0.222559	0.859795	1.116463

- Confidence: 88.9% of those who bought pasta and meatballs also bought oil
- Support: This rule appears in 15% of all transactions in the cluster 
- Lift:The likelihood of buying oil increases by 1.2 if pasta and meatballs are bought.

Cluster 8:
(spaghetti)	(fresh tuna, champagne)	0.335789	0.448065	0.180134	0.536450	1.19725

- Confidence: 88.9% of those who bought pasta and meatballs also bought oil
- Support: This rule appears in 15% of all transactions in the cluster 
- Lift:The likelihood of buying oil increases by 1.2 if pasta and meatballs are bought.




## Eclat Algorithm

In [39]:
items_list2

[['soup',
  'shallot',
  'cake',
  'fresh bread',
  'cream',
  'carrots',
  'babies food',
  'ham',
  'cookies',
  'salad',
  'tomato sauce',
  'cooking oil',
  'tea'],
 ['candy bars', 'soup', 'napkins', 'rice', 'oil', 'tea'],
 ['zucchini',
  'grated cheese',
  'ketchup',
  'oil',
  'pokemon scarlet',
  'milk',
  'water spray',
  'chicken',
  'cooking oil',
  'tea'],
 ['candy bars',
  'napkins',
  'frozen smoothie',
  'oil',
  'ham',
  'phone car charger',
  'cologne',
  'iMac',
  'tea'],
 ['cake', 'cream', 'corn', 'blueberries', 'burgers'],
 ['oil', 'babies food', 'cooking oil'],
 ['soup', 'protein bar', 'oil', 'toothpaste', 'ham', 'cooking oil', 'tea'],
 ['candy bars',
  'cake',
  'whole wheat rice',
  'oil',
  'portal',
  'black tea',
  'fresh bread',
  'ham',
  'green beans',
  'cooking oil'],
 ['meatballs',
  'cottage cheese',
  'french fries',
  'turkey',
  'oil',
  'chocolate bread',
  'tomatoes',
  'cologne',
  'cooking oil',
  'tea'],
 ['soup',
  'cooking oil',
  'cereals',
  

In [40]:
eclat_algorithm(items_list2, min_combination = 2, max_combination = 2, min_support = 0.02)

Combination 2 by 2


6786it [01:14, 91.36it/s] 


,support
oil & cooking oil,0.428647
tea & oil,0.426828
cake & oil,0.329480
tea & cooking oil,0.304542
candy bars & oil,0.261237
...,...
napkins & rice,0.020185
shallot & oil,0.020185
tea & deodorant,0.020185
babies food & fresh bread,0.020127
